09-spring-boot-ehcache-xml/src/main/java/com/example/springbootehcache/ext/ehcache/EhcachejCacheSpringConfiguration.java

```diff
@@ -0,0 +1,117 @@
+import org.springframework.boot.autoconfigure.cache.JCacheManagerCustomizer;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.annotation.Bean;
+import org.springframework.core.io.Resource;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.StringUtils;
+
+import javax.cache.CacheManager;
+import javax.cache.Caching;
+import javax.cache.configuration.MutableConfiguration;
+import javax.cache.spi.CachingProvider;
+import java.io.IOException;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Properties;
+
+@AutoConfiguration
+@ConditionalOnClass({org.ehcache.CacheManager.class})
+@ConditionalOnMissingBean(value = org.springframework.cache.CacheManager.class, name = "cacheResolver")
+@EnableConfigurationProperties(CacheProperties.class)
+@ConditionalOnProperty("spring.cache.jcache.config")
+public class EhcachejCacheSpringConfiguration implements BeanClassLoaderAware, BeanFactoryAware {
+  private ClassLoader beanClassLoader;
+  private BeanFactory beanFactory;
+
+  @Override
+  public void setBeanClassLoader(ClassLoader classLoader) {
+    this.beanClassLoader = classLoader;
+  }
+
+  @Override
+  public void setBeanFactory(BeanFactory beanFactory) throws BeansException {
+    this.beanFactory = beanFactory;
+  }
+
+  @Bean
+  @ConditionalOnMissingBean
+  public InstantiatorService instantiatorService() {
+    return new DefaultInstantiatorService();
+  }
+
+  @Bean
+  public javax.cache.CacheManager jCacheCacheManager(CacheProperties cacheProperties,
+                                                     ObjectProvider<javax.cache.configuration.Configuration<?, ?>> defaultCacheConfiguration,
+                                                     ObjectProvider<JCacheManagerCustomizer> cacheManagerCustomizers,
+                                                     InstantiatorService delegate) throws IOException {
+    CacheManager jCacheCacheManager = createCacheManager(cacheProperties, delegate);
+    List<String> cacheNames = cacheProperties.getCacheNames();
+    if (!CollectionUtils.isEmpty(cacheNames)) {
+      for (String cacheName : cacheNames) {
+        jCacheCacheManager.createCache(cacheName,
+            defaultCacheConfiguration.getIfAvailable(MutableConfiguration::new));
+      }
+    }
+    cacheManagerCustomizers.orderedStream().forEach((customizer) -> customizer.customize(jCacheCacheManager));
+    return jCacheCacheManager;
+  }
+
+  private CacheManager createCacheManager(CacheProperties cacheProperties, InstantiatorService delegate) throws IOException {
+    CachingProvider cachingProvider = getCachingProvider(cacheProperties.getJcache().getProvider());
```

Comment:
If we're in "Ehcache-land" at this point (or we know we're not and can avoid any more work), why not just cast the `CachingProvider` to a implementation specific type and pass the services that way? Then we could do things in a much safer way rather than having to pass things through a properties object?

Reply:
We are not: here, we do exactly like the spring boot code is doing, which is creating a jsr `CacheManager` based on a potentially provided and customised `CacheProperties`, and then customise the created jsr CacheManager.

The only Ehcache related thing is setting this `ehcache.service.instantiatorService` for the JSR provider implementation.

Spring is providing the config URI and jsr properties. And we cannot call directly Ehcache implementation `EhcacheCachingProvider.getCacheManager(uri, configuration, properties)` because it would means we have to duplicate the Ehcahe internal code to parle the configuration from a URI or provide a default one (`ConfigSupplier` code). And the `ConfigSupplier` class is not even publicly usable.